lib/src/multipart_request.dart

```diff
@@ -4,162 +4,143 @@
-    throw new UnsupportedError("Cannot set the contentLength property of "
-        "multipart requests.");
-  }
-
-  /// Freezes all mutable fields and returns a single-subscription [ByteStream]
-  /// that will emit the request body.
-  ByteStream finalize() {
-    // TODO(nweiz): freeze fields and files
-    var boundary = _boundaryString();
-    headers['content-type'] = 'multipart/form-data; boundary=$boundary';
-    super.finalize();
-
+class MultipartBody implements Body {
+  /// The contents of the message body.
+  ///
+  /// This will be `null` after [read] is called.
+  Stream<List<int>> _stream;
+
+  /// The length of the stream returned by [read].
+  ///
+  /// This is calculated from the fields and files passed into the body.
+  final int contentLength;
+
+  /// Multipart forms do not have an encoding.
+  Encoding get encoding => null;
+
+  /// Creates a [MultipartBody] from the given [fields] and [files].
+  ///
+  /// The [boundary] is used to
+  factory MultipartBody(
+    Map<String, String> fields,
+    List<MultipartFile> files,
+    String boundary,
+  ) {
     var controller = new StreamController<List<int>>(sync: true);
+    var contentLength = 0;
 
     void writeAscii(String string) {
-      controller.add(UTF8.encode(string));
+      controller.add(string.codeUnits);
+      contentLength += string.length;
+    }
+    void writeUtf8(String string) {
+      var encoded = UTF8.encode(string);
+      controller.add(encoded);
+      contentLength += encoded.length;
+    }
+    void writeLine() {
+      controller.add([13, 10]); // \r\n
+      contentLength += 2;
     }
 
-    writeUtf8(String string) => controller.add(UTF8.encode(string));
-    writeLine() => controller.add([13, 10]); // \r\n
-
+    // Write the fields to the stream
     fields.forEach((name, value) {
       writeAscii('--$boundary\r\n');
-      writeAscii(_headerForField(name, value));
+      writeUtf8(_headerForField(name, value));
       writeUtf8(value);
       writeLine();
     });
 
-    Future.forEach(_files, (file) {
-      writeAscii('--$boundary\r\n');
-      writeAscii(_headerForFile(file));
-      return writeStreamToSink(file.finalize(), controller)
-        .then((_) => writeLine());
+    // Iterate over the files to get the length
+    var fileCount = files.length;
+    var fileHeaders = new List<List<int>>(fileCount);
+
+    for (var i = 0; i < fileCount; ++i) {
+      var file = files[i];
+      var header = <int>[];
+      header.addAll('--$boundary\r\n'.codeUnits);
+      header.addAll(UTF8.encode(_headerForFile(file)));
+      contentLength += header.length + file.length + 2;
+      fileHeaders[i] = header;
+    }
+
+    // Ending characters
+    var ending = '--$boundary--\r\n'.codeUnits;
+    contentLength += ending.length;
+
+    // Write the files to the stream
+    //
+    // Future.forEach will ensure that the actions happen in sequence so i is
+    // used to get the fileHeaders
+    var i = 0;
+
+    Future.forEach(files, (file) {
```

Comment:
Consider pulling this into a static method so you can make it async and just use a normal `for` loop.